src/renderer-backend-egl.cpp

```diff
@@ -96,139 +116,161 @@ class Backend {
-        g_source_set_name(m_source, "WPEBackend-fdo::Backend");
-        g_source_set_priority(m_source, -70);
-        g_source_set_can_recurse(m_source, TRUE);
-        g_source_attach(m_source, g_main_context_get_thread_default());
-
-        m_registry = wl_display_get_registry(m_display);
-        wl_registry_add_listener(m_registry, &s_registryListener, this);
-        wl_display_roundtrip(m_display);
     }
 
-    ~Backend()
-    {
-        if (m_source) {
-            g_source_destroy(m_source);
-            g_source_unref(m_source);
-        }
-    }
+    ~Backend() = default;
 
     struct wl_display* display() const { return m_display; }
-    struct wl_compositor* compositor() const { return m_compositor; }
 
 private:
-    static const struct wl_registry_listener s_registryListener;
-
     struct wl_display* m_display;
-    struct wl_registry* m_registry;
-    struct wl_compositor* m_compositor;
-
-    GSource* m_source;
-};
-
-const struct wl_registry_listener Backend::s_registryListener = {
-    // global
-    [](void* data, struct wl_registry* registry, uint32_t name, const char* interface, uint32_t)
-    {
-        auto& backend = *reinterpret_cast<Backend*>(data);
-
-        if (!std::strcmp(interface, "wl_compositor"))
-            backend.m_compositor = static_cast<struct wl_compositor*>(wl_registry_bind(registry, name, &wl_compositor_interface, 1));
-    },
-    // global_remove
-    [](void*, struct wl_registry*, uint32_t) { },
 };
 
 class Target {
 public:
     Target(struct wpe_renderer_backend_egl_target* target, int hostFd)
         : m_target(target)
     {
-        m_socket = g_socket_new_from_fd(hostFd, nullptr);
-        if (m_socket)
-            g_socket_set_blocking(m_socket, FALSE);
+        m_glib.socket = g_socket_new_from_fd(hostFd, nullptr);
+        if (m_glib.socket)
+            g_socket_set_blocking(m_glib.socket, FALSE);
     }
 
     ~Target()
     {
-        if (m_frameCallback)
-            wl_callback_destroy(m_frameCallback);
-        if (m_window)
-            wl_egl_window_destroy(m_window);
-        if (m_surface)
-            wl_surface_destroy(m_surface);
-
-        if (m_socket)
-            g_object_unref(m_socket);
-        if (m_source) {
-            g_source_destroy(m_source);
-            g_source_unref(m_source);
+        if (m_wl.frameCallback)
+            wl_callback_destroy(m_wl.frameCallback);
+        if (m_wl.window)
+            wl_egl_window_destroy(m_wl.window);
+        if (m_wl.surface)
+            wl_surface_destroy(m_wl.surface);
+
+        if (m_wl.compositor)
+            wl_compositor_destroy(m_wl.compositor);
+        if (m_wl.registry)
+            wl_registry_destroy(m_wl.registry);
+        if (m_wl.eventQueue)
+            wl_event_queue_destroy(m_wl.eventQueue);
+        if (m_wl.displayWrapper)
+            wl_proxy_wrapper_destroy(m_wl.displayWrapper);
+
+        if (m_glib.socket)
+            g_object_unref(m_glib.socket);
```

Comment:
I would just replace all the manual `NULL`-checking with `g_clear_pointer()` and `g_clear_object()`, e.g:

```c
g_clear_pointer (&m_wl.compositor, wl_compositor_destroy);
g_clear_pointer (&m_wl.registry, wl_registry_destroy);
// ...
g_clear_object (&m_glib.socket);
// ...
```

On top of making the code more concise, using these ensures that the pointers are always set to `NULL` which is handy to help find out use-after-free situations.